alerts/machine-config-operator/MachineConfigControllerDrainError.md

```diff
@@ -0,0 +1,114 @@
+```
+
+If the drain continues to fail then another failure log will fire.
+
+```console
+ 1 drain_controller.go:173] node xxxxxxx-xxxxxx-xxxxxx-xxxxx: Drain failed. Drain has been failing for more than 10 minutes. Waiting 5 minutes then retrying. Error message from drain: error when waiting for pod "xxxx-xxxx-xx" in namespace "xxxxxxx" to terminate: global timeout reached: 1m30s
+```
+
+Finally if 1 hour has passed and the drain is still failing the following logs will be fired and the node will be marked degraded.
+
+```console
+1 drain_controller.go:352] node xxxxxxx-xxxxxx-xxxxxx-xxxxx: drain exceeded timeout: 1h0m0s. Will continue to retry.
+```
+
+```console
+1 status.go:126] Degraded Machine: xxxxxxx-xxxxxx-xxxxxx-xxxxx: and Degraded Reason: failed to drain node: xxxxxxx-xxxxxx-xxxxxx-xxxxx after 1 hour. Please see machine-config-controller logs for more information
+```
+
+The machine-controller logs will explicitly tell you what pods are failing to drain succesfully. Looking at the pods named in the logs is where you should start. 
+
+Common reasons a node cannot drain a pod successfully are:
+
+- The pod has a PodDisruptionBudget (PBD) set preventing the controller from deleting the pod.
+- The pod has storage attached and the kubelet is unable to unmount the storage succesfully.
+- The pod deletion may be prevented by a webhook if it's configured to target UPDATE operations or the webhook is failing to be called by the kube-apiserver.
+- The pod deletion could be stuck because the finalizers set in the pod are preventing it from terminating succesfully.
+
+## Mitigation 
+
+- If a PDB is causing the failure you can temporarily patch it to set the `minAvailable` to 0 so it can scale the pod down successfully. Then patch it to the old value after the upgrade completes.
+
+```console
+oc patch pdb $PDB -n $NS --type=merge -p '{"spec":{"minAvailable":0}}'
+```
+- If a webhook is preventing the pod deletion you can check the `openshift-kube-apiserver` logs to see exactly what webhook is preventing deletion.
+
+```console
+Failed calling webhook, failing open vault.hashicorp.com: failed calling webhook "webhook_name": failed to call webhook: Post "https://hashi-vault-agent-injector-svc.vault-injector.svc:443/mutate?timeout=30s": context canceled2022-10-11T19:13:53.345381984Z E1011 19:13:53.345348      16 dispatcher.go:184] failed calling webhook "webhook_name": failed to call webhook: Post "https://name-injector-svc.vault-injector.svc:443/mutate?timeout=30s": context canceled
+```
+Then check if it's a `mutating` or `validating` webhook.
+```console
+$ oc get validatingwebhookconfiguration
+$ oc get mutatingwebhookconfiguration
+```
+Then backup and delete the webhook 
+```console
+$ oc get validatingwebhookconfiguration/<webhook_name> -o yaml > webhook.yaml
+$$ oc delete <webhook_type> <webhook_name>         
+```
+This should allow the drain to continue and if the webhook does not come back automatically you can recreate it via the backup.
```

Comment:
I can't get the formatting to look right in the comment preview, but you get the idea:
```suggestion
- If a webhook is preventing the pod deletion:
+
. Check the `openshift-kube-apiserver` logs to see exactly what webhook is preventing deletion.
+
```console
Failed calling webhook, failing open vault.hashicorp.com: failed calling webhook "webhook_name": failed to call webhook: Post "https://hashi-vault-agent-injector-svc.vault-injector.svc:443/mutate?timeout=30s": context canceled2022-10-11T19:13:53.345381984Z E1011 19:13:53.345348      16 dispatcher.go:184] failed calling webhook "webhook_name": failed to call webhook: Post "https://name-injector-svc.vault-injector.svc:443/mutate?timeout=30s": context canceled
```
+
. Check if it's a `mutating` or `validating` webhook.
+
```console
$ oc get validatingwebhookconfiguration
$ oc get mutatingwebhookconfiguration
```
+
. Backup and delete the webhook: 
+
```console
$ oc get validatingwebhookconfiguration/<webhook_name> -o yaml > webhook.yaml
$$ oc delete <webhook_type> <webhook_name>         
```
+
This should allow the drain to continue and if the webhook does not come back automatically you can recreate it via the backup.
```

Reply:
Agreed, Applied the change